lib/sync/fullsync.go

```diff
@@ -0,0 +1,547 @@
+	if cfg.NumOfTasks == 0 {
+		cfg.NumOfTasks = defaultNumOfTasks
+	}
+
+	return &FullSyncStrategy{
+		badBlocks:  cfg.BadBlocks,
+		reqMaker:   cfg.RequestMaker,
+		blockState: cfg.BlockState,
+		numOfTasks: cfg.NumOfTasks,
+		importer:   newBlockImporter(cfg),
+		unreadyBlocks: &unreadyBlocks{
+			incompleteBlocks: make(map[common.Hash]*types.BlockData),
+			// TODO: cap disjoitChains to don't grows indefinitely
+			disjointChains: make([][]*types.BlockData, 0),
+		},
+		requestQueue: &requestsQueue[*network.BlockRequestMessage]{
+			queue: list.New(),
+		},
+		peers: &peerViewSet{
+			view:   make(map[peer.ID]peerView),
+			target: 0,
+		},
+	}
+}
+
+func (f *FullSyncStrategy) NextActions() ([]*syncTask, error) {
+	f.startedAt = time.Now()
+	f.syncedBlocks = 0
+
+	if f.requestQueue.Len() > 0 {
+		message, _ := f.requestQueue.PopFront()
+		return f.createTasks([]*network.BlockRequestMessage{message}), nil
+	}
+
+	currentTarget := f.peers.getTarget()
+	bestBlockHeader, err := f.blockState.BestBlockHeader()
+	if err != nil {
+		return nil, fmt.Errorf("while getting best block header")
+	}
+
+	// our best block is equal or ahead of current target.
+	// in the nodes pov we are not legging behind so there's nothing to do
+	if uint32(bestBlockHeader.Number) >= currentTarget {
+		return nil, nil
+	}
+
+	startRequestAt := bestBlockHeader.Number + 1
+	targetBlockNumber := startRequestAt + 128
+
+	if targetBlockNumber > uint(currentTarget) {
+		targetBlockNumber = uint(currentTarget)
+	}
+
+	ascendingBlockRequests := network.NewAscendingBlockRequests(startRequestAt, targetBlockNumber,
+		network.BootstrapRequestData)
+
+	return f.createTasks(ascendingBlockRequests), nil
+}
+
+func (f *FullSyncStrategy) createTasks(requests []*network.BlockRequestMessage) []*syncTask {
+	tasks := make([]*syncTask, len(requests))
+	for idx, req := range requests {
+		tasks[idx] = &syncTask{
+			request:      req,
+			response:     &network.BlockResponseMessage{},
+			requestMaker: f.reqMaker,
+		}
+	}
+	return tasks
+}
+
+func (f *FullSyncStrategy) IsFinished(results []*syncTaskResult) (bool, []Change, []peer.ID, error) {
+	repChanges, peersToIgnore, validResp := validateResults(results, f.badBlocks)
+
+	validBlocksUnderFragment := func(highestFinalizedNumber uint, fragmentBlocks []*types.BlockData) []*types.BlockData {
+		startFragmentFrom := -1
+		for idx, block := range fragmentBlocks {
+			if block.Header.Number > highestFinalizedNumber {
+				startFragmentFrom = idx
+				break
+			}
+		}
+
+		if startFragmentFrom < 0 {
+			return nil
+		}
+
+		return fragmentBlocks[startFragmentFrom:]
+	}
+
+	highestFinalized, err := f.blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return false, nil, nil, fmt.Errorf("getting highest finalized header")
+	}
+
+	readyBlocks := make([][]*types.BlockData, 0, len(validResp))
+	for _, reqRespData := range validResp {
+		// if Gossamer requested the header, then the response data should
+		// contains the full bocks to be imported
```

Comment:
```suggestion
		// contains the full block to be imported
```

Reply:
not sure if this is one or multiple blocks

Reply:
can be both, will keep in the plural.